### PR TITLE
feat(config): support custom global bin dir

### PR DESCRIPTION
**Pull request by umireon**
## Abstract
It would be great that the package manager can customize the global bin directory along with its policy.
I propose `global-bin-dir` in a npmrc to customize the global bin directory of pnpm.

## Use case
https://github.com/Homebrew/homebrew-core/pull/82892
In Homebrew, create `/usr/local/etc/npmrc` and add `global-bin-dir = ${HOME}/Library/pnpm` to it.
It will change the global bin directory to `~/Library/pnpm` without PNPM_HOME set.

(In Linuxbrew, `~/.local/pnpm` would be used)

## References
#3713

### CODE REVIEW

**Comment by umireon** [comment]
This change is for https://github.com/Homebrew/homebrew-core/pull/82892/commits/4546e4c0cbcccb51bd87ea12ec62163d6268be22.

**Comment by zkochan** [comment]
pnpm uses npm's config system and we already read the global npmrc file. Can't we just reuse that?

**Comment by umireon** [comment]
@zkochan I removed pnpmrc and this PR is just adding `pnpm-bin` option to specify the global bin dir.

**Comment by umireon** [comment]
`pkg -t macos-node14-arm64` must fail on Intel Mac so that we need to fix `build-artifacts.ts` in another PR.

File: packages/config/src/index.ts [hunk 1]
```diff
@@ -71,6 +71,7 @@ export const types = Object.assign({
   offline: Boolean,
   'package-import-method': ['auto', 'hardlink', 'clone', 'copy'],
   pnpmfile: String,
+  'pnpm-bin': String,
```
Comment (zkochan) [inline packages/config/src/index.ts]: might be better to name it `global-bin-dir`

Comment (umireon) [reply packages/config/src/index.ts]: Changed.

File: packages/config/src/index.ts [hunk 2]
```diff
@@ -275,13 +276,18 @@ export default async (
...
+    }
+
     pnpmConfig.bin = cliOptions.dir
       ? (
         process.platform === 'win32'
           ? cliOptions.dir
           : path.resolve(cliOptions.dir, 'bin')
       )
-      : globalBinDir(knownGlobalBinDirCandidates, { shouldAllowWrite: opts.globalDirShouldAllowWrite === true })
+      : npmConfigGlobalBinDir || globalBinDir(knownGlobalBinDirCandidates, { shouldAllowWrite: opts.globalDirShouldAllowWrite === true })
```
Comment (zkochan) [inline packages/config/src/index.ts]: I think if `global-bin-dir` is specified, it should be preferred regardless of system and/or other settings.

Comment (umireon) [reply packages/config/src/index.ts]: Do you mean cliOptions.dir should not override `global-bin-dir` in a npmrc?

Comment (umireon) [reply packages/config/src/index.ts]: `globalBinDir` won't be called if npmConfig.get('global-bin-dir') is truthy.

File: packages/config/src/index.ts [hunk 3]
```diff
@@ -275,13 +276,22 @@ export default async (
...
+    }
+
     pnpmConfig.bin = cliOptions.dir
       ? (
         process.platform === 'win32'
           ? cliOptions.dir
           : path.resolve(cliOptions.dir, 'bin')
       )
-      : globalBinDir(knownGlobalBinDirCandidates, { shouldAllowWrite: opts.globalDirShouldAllowWrite === true })
+      : preferredGlobalBinDir
```
Comment (zkochan) [inline packages/config/src/index.ts]: I  mean

```ts
    const npmConfigGlobalBinDir = npmConfig.get('global-bin-dir')
    if (typeof npmConfigGlobalBinDir === 'string') {
      fs.mkdirSync(npmConfigGlobalBinDir, { recursive: true })
      pnpmConfig.bin = npmConfigGlobalBinDir
    } else {
      pnpmConfig.bin = cliOptions.dir
        ? (
          process.platform === 'win32'
            ? cliOptions.dir
            : path.resolve(cliOptions.dir, 'bin')
        )
        : globalBinDir(knownGlobalBinDirCandidates, { shouldAllowWrite: opts.globalDirShouldAllowWrite === true })
    }
```

Comment (umireon) [reply packages/config/src/index.ts]: Sure

**Comment by zkochan** [comment]
some tests should be added.

File: packages/config/test/global-bin-dir.test.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/// <reference path="../../../typings/index.d.ts"/>
+import path from 'path'
+import { homedir } from 'os'
+import getConfig from '@pnpm/config'
+
```
Comment (zkochan) [inline packages/config/test/global-bin-dir.test.ts]: use `globalBinDir.test.ts` to name this file

**Comment by umireon** [comment]
I'm struggling to create tests😅

File: packages/config/test/global-bin-dir.test.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  const { config } = await getConfig({
+    cliOptions: {
+      global: true,
+    },
+    packageManager: {
+      name: 'pnpm',
+      version: '1.0.0',
+    },
+  })
+  expect(config.bin).toBe(path.join(homedir(), 'Library', 'pnpm'))
```
Comment (zkochan) [inline packages/config/test/global-bin-dir.test.ts]: I assume this test will fail on Windows.

Comment (umireon) [reply packages/config/test/global-bin-dir.test.ts]: Observe whether this is fixed.

**Review by zkochan — APPROVED** [review]

**Review by zkochan — COMMENTED** [review]
Please add a changeset for your change

Run `pnpm changeset`

**Comment by umireon** [comment]
@zkochan
> Please add a changeset for your change

Yes, tomorrow. I have to wait for CI.

**Comment by umireon** [comment]
slash and backslash problem.

**Comment by umireon** [comment]
Investigating

**Comment by umireon** [comment]
`os.tmpdir()` returns `/var/folders/d5/*/T` but `/var` is symlinked to `/private/var` so that the realpath is `/private/var/folders/d5/*/T` and it causes the error only in macOS.

**Comment by umireon** [comment]
Oh thanks for the changeset. I couldn't wait for CI and fell asleep.

**Pull request merged**